c2/c2-client-bundle/c2-client-http/src/main/java/org/apache/nifi/c2/client/http/url/ProxyAwareC2UrlProvider.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nifi.c2.client.http.url;
+
+import static org.apache.commons.lang3.StringUtils.isNotBlank;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+
+public class ProxyAwareC2UrlProvider implements C2UrlProvider {
+
+    private final String c2RestApi;
+    private final String c2RestPathHeartbeat;
+    private final String c2RestPathAcknowledge;
+
+    ProxyAwareC2UrlProvider(String c2RestApi, String c2RestPathHeartbeat, String c2RestPathAcknowledge) {
+        this.c2RestApi = c2RestApi;
+        this.c2RestPathHeartbeat = c2RestPathHeartbeat;
+        this.c2RestPathAcknowledge = c2RestPathAcknowledge;
+    }
+
+    @Override
+    public String getHeartbeatUrl() {
+        return toAbsoluteUrl(c2RestPathHeartbeat);
+    }
+
+    @Override
+    public String getAcknowledgeUrl() {
+        return toAbsoluteUrl(c2RestPathAcknowledge);
+    }
+
+    @Override
+    public String getCallbackUrl(String absoluteUrl, String relativeUrl) throws Exception {
```

Comment:
What do you think about making the decisions below in C2HttpClient (where this is called) and introducing a "fromRelativeUrl" and a "fromAbsoluteUrl" function and the fromAbsolute could have  adefault implementation in C2UrlProvider.

Reply:
With the refactor I think it became clearer. I would leave the logic as it is, as the client should be unaware how to create the callback url, that logic should be encapsulated in the UrlProvider classes.
The reason it is exposed through C2Client, is that I didn't want the UrlProvider classes to be accessed from the OperationHandlers. Hence C2Client already contained the UrlProvider, it seemed convenient to implement this way.